# Genetic Simulations Tutorials - README

## Project Overview

Welcome to the **Genetic Simulations Tutorials** repository :wave:

This project is a collection of R tutorials designed to guide you through various aspects of simulating genetic data, focusing on Single Nucleotide Polymorphisms (SNPs), Linkage Disequilibrium (LD) and espistasis. The tutorials are written in R Markdown (Rmd) format for now, but will ultimately be in quarto and available in both Python and R

**Author:** Matthew Bracher-Smith  
**Date Created:** 25/01/2024

---

## Current Progress

**Status:** Work in Progress

This repository is actively being developed. New tutorials and updates will be added as the project evolves. As of now, the tutorials cover:

- Simulating unassociated SNPs
- Additive liability-threshold models
- Multiplicative interaction models
- Generating simple LD
- Creating SNP data with varying LD patterns
- Simulating complex LD blocks

Each tutorial will include detailed explanations and R code snippets to demonstrate the concepts.

---

## Usage

To use these tutorials, you will need to have R installed on your machine along with the following packages:

- `dplyr`
- `ggcorrplot`
- Other dependencies may be required as the project grows

The tutorials are structured as R Markdown files. To view or run the tutorials, open the Rmd files in an R environment such as RStudio.

---

## Requirements

R:

- tidyverse

Other:

- [bcftools](https://www.htslib.org/download/) (Only needed for sampling from 1000 genomes remotely)

---

## Contributing

As this is a work in progress, contributions, suggestions, and feedback are highly appreciated. If you have ideas for additional topics or improvements, please feel free to submit an issue or pull request.

---

## License

This project is open-sourced under the GPL-3.0 license. See the LICENSE file for more details.

---

**Note:** This README is a living document and will be updated regularly to reflect the latest state of the project. Stay tuned for more!

---
